pkg/console/service/service.go

```diff
@@ -0,0 +1,353 @@
+package service
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/emicklei/go-restful/v3"
+	authnv1 "k8s.io/api/authentication/v1"
+	authzv1 "k8s.io/api/authorization/v1"
+	core "k8s.io/api/core/v1"
+	rbac "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+	kubevirt "kubevirt.io/api/core"
+	kubevirtv1 "kubevirt.io/api/core/v1"
+	"kubevirt.io/client-go/kubecli"
+
+	"github.com/kubevirt/vm-console-proxy/api/v1alpha1"
+)
+
+const (
+	AppKubernetesNameLabel      = "app.kubernetes.io/name"
+	AppKubernetesPartOfLabel    = "app.kubernetes.io/part-of"
+	AppKubernetesVersionLabel   = "app.kubernetes.io/version"
+	AppKubernetesManagedByLabel = "app.kubernetes.io/managed-by"
+	AppKubernetesComponentLabel = "app.kubernetes.io/component"
+)
+
+type Service interface {
+	TokenHandler(request *restful.Request, response *restful.Response)
+	VncHandler(request *restful.Request, response *restful.Response)
+}
+
+func NewService(kubevirtClient kubecli.KubevirtClient) Service {
+	return &service{
+		kubevirtClient: kubevirtClient,
+	}
+}
+
+type service struct {
+	kubevirtClient kubecli.KubevirtClient
+}
+
+func (s *service) TokenHandler(request *restful.Request, response *restful.Response) {
+	params, err := readTokenRequestParameters(request)
+	if err != nil {
+		_ = response.WriteError(http.StatusBadRequest, err)
+		return
+	}
+
+	authToken := getAuthToken(request)
+	if authToken == "" {
+		_ = response.WriteError(http.StatusUnauthorized, fmt.Errorf("authenticating token cannot be empty"))
```

Comment:
IMO we should not give more info on 401. We could potentially log internally, but over the API I would not return more information to a potential attacker than necessary.